pkgs/by-name/sp/spectral-language-server/package.nix

```
@@ -0,0 +1,107 @@
+{ lib
+, buildNpmPackage
+, mkYarnPackage
+, fetchYarnDeps
+, fetchFromGitHub
+, typescript
+, jq
+, fetchpatch
+}:
+let
+  # Instead of the build script that spectral-language-server provides (ref: https://github.com/luizcorreia/spectral-language-server/blob/master/script/vscode-spectral-build.sh), we build vscode-spectral manually.
+  # This is because the script must go through the network and will not work under the Nix sandbox environment.
+  vscodeSpectral = mkYarnPackage rec {
+    pname = "vscode-spectral";
+    version = "1.1.2";
+
+    src = fetchFromGitHub {
+      owner = "stoplightio";
+      repo = "vscode-spectral";
+      rev = "v${version}";
+      hash = "sha256-TWy+bC6qhTKDY874ORTBbvCIH8ycpmBiU8GLYxBIiAs=";
+    };
+
```

```suggestion

  packageJSON = ./package.json;

```
The ofborg error seems to be caused by [IFD (Import From Derivation)](https://nixos.org/manual/nix/stable/language/import-from-derivation).
Looking at other packages that use `mkYarnPackage`, they seem to get around this by committing upstream package.json to nixpkgs.

---

I could not find any causes for ofborg throwing errors, so this was helpful. I have added the package.json in vscode-spectral to nixpkgs.
Sorry to bother you, I'd like to confirm my understanding. Does `mkYarnPackage` need package.json because it tries to fetch dependencies with IFD before downloading from GitHub?

---

IFD seems to be caused not by fetching dependencies but by calling `importJSON`. `importJSON` executes `readFile` internally.
https://github.com/NixOS/nixpkgs/blob/82cd9ea4846b0472dc574b158afac35014c19372/pkgs/development/tools/yarn2nix-moretea/yarn2nix/default.nix#L269

https://github.com/NixOS/nixpkgs/blob/82cd9ea4846b0472dc574b158afac35014c19372/lib/trivial.nix#L639-L640

To be honest, I had no idea what the ofborg error said. 🤷

---

I see. I don't understand it perfectly, but I will try to read `mkYarnPackage` and other various derivations. Thank you very much.

---

FYI: https://github.com/NixOS/nixpkgs/issues/296856
It explained why `packageJSON` would cause an IFD.
And it seems that we should not have used mkYarnPackage in the first place. 🫠